File: ironclient/lib/models/hostedfurnace.dart
import 'package:ironcirclesapp/models/hostedfurnaceimage.dart';

class HostedFurnace {
  String name;
  String key;
  String id;
  HostedFurnaceImage? hostedFurnaceImage;
  String description;
  String link;
  bool adultOnly;
  bool memberAutonomy;
  bool override;
  bool approved;
  bool discoverable;
  bool enableWall;

  HostedFurnace({
    required this.name,
    required this.key,
    this.hostedFurnaceImage,
    required this.id,
    required this.description,
    required this.link,
    required this.adultOnly,
    required this.memberAutonomy,
    required this.override,
    required this.approved,
    required this.discoverable,
    required this.enableWall,
  });

  factory HostedFurnace.fromJson(Map<String, dynamic> jsonMap) => HostedFurnace(
        name: jsonMap['name'],
        key: jsonMap['key'] ?? '',
        id: jsonMap['_id'],
        hostedFurnaceImage: jsonMap["hostedFurnaceImage"] == null
            ? null
            : HostedFurnaceImage.fromJson(jsonMap["hostedFurnaceImage"]),
        description:
            jsonMap["description"] == null ? '' : jsonMap['description'],
        link:
            jsonMap["link"] == null ? '' : jsonMap["link"],
        adultOnly: jsonMap["adultOnly"],
        memberAutonomy: jsonMap["memberAutonomy"] ?? true,
        override: jsonMap["override"]== null ? false : jsonMap["override"],
        approved: jsonMap["approved"] == null ? false : jsonMap["approved"],
        discoverable: jsonMap["discoverable"],
        enableWall: jsonMap["enableWall"],
      );

  Map<String, dynamic> toJson() => {
        'name': name,
        'key': key,
        '_id': id,
        'hostedFurnaceImage':
            hostedFurnaceImage == null ? null : hostedFurnaceImage!.toJson(),
        'description': description,
        'link': link,
        'adultOnly': adultOnly ? true : false,
        'memberAutonomy': memberAutonomy,
        'override': override ? true : false,
        'approved': approved ? true : false,
        'discoverable': discoverable,
        'enableWall': enableWall,
      };
}

class HostedFurnaceCollection {
  final List<HostedFurnace> objects;

  HostedFurnaceCollection.fromJSON(Map<String, dynamic> json, String key)
      : objects = (json[key] as List)
            .map((json) => HostedFurnace.fromJson(json))
            .toList();
}

File: ironapi/models/hostedfurnace.js
const mongoose = require('mongoose');

var HostedFurnaceStorage = require('../models/hostedfurnacestorage');
var HostedFurnaceImage = require('../models/hostedfurnaceimage');

var HostedFurnaceSchema = new mongoose.Schema({
  name: { type: String, required: true },
  lowercase: { type: String, required: true },
  key: { type: String, required: true },
  discoverable: { type: Boolean, default: false },
  approved: { type: Boolean, default: false }, ///networks are not discoverable until approved by admin
  storeApproved: { type: Boolean }, ///networks are not discoverable until approved by admin, no default, so it is not set until approved
  override: { type: Boolean, default: false }, ///ability to change a network to discoverable is blocked, used to disable discovery for a network, for example if hate speech is found
  enableWall: { type: Boolean, default: false },
  wallCircleID: { type: String, }, //not an object ref on purpose, would cause a circular reference
  adultOnly: { type: Boolean, default: false },
  memberAutonomy: { type: Boolean, default: true }, ///can members create circles/vaults & send invites
  type: { type: Number, default: 1 },
  description: { type: String },
  link: { type: String },
  hostedFurnaceImage: mongoose.model('HostedFurnaceImage').schema,
  storage: [mongoose.model('HostedFurnaceStorage').schema],
}, { timestamps: { createdAt: 'created', updatedAt: 'lastUpdate' } }, { collection: 'hostedfurnace' });

HostedFurnaceSchema.statics.new = async function (json) {

  let object = this(json);
  return object;
}


mongoose.model('HostedFurnace', HostedFurnaceSchema);

module.exports = mongoose.model('HostedFurnace');